specs/experimental/custom-gas-token.md

```diff
@@ -0,0 +1,306 @@
+    address _to,
+    uint256 _mint,
+    uint256 _value,
+    uint64 _gasLimit,
+    bool _isCreation,
+    bytes memory _data
+) public;
+```
+
+#### `depositTransaction`
+
+The `depositTransaction` function is updated to revert if there is non-zero `msg.value` when the system uses a custom
+native token. This allows existing applications that do cross domain calls to continue functioning as is.
+
+### StandardBridge
+
+The `StandardBridge` contracts on both L1 and L2 MUST be aware of the custom gas token. The `ether` specific ABI on the
+`StandardBridge` is disabled when custom gas token is enabled.
+
+This includes the following methods:
+
+- `bridgeETH(uint32,bytes)`
+- `bridgeETHTo(address,uint32,bytes)`
+- `receive()`
+
+The following diagram shows the control flow for when a user attempts to send `ether` through the `StandardBridge`, either
+on L1 or L2.
+
+![](../static/assets/custom-token-bridge-revert.png)
+
+### CrossDomainMessenger
+
+The `CrossDomainMessenger` contracts on both L1 and L2 MUST not be used to send the native asset between domains.
+It is possible to support this in the future, but for now the feature is disabled due to the high probability of footguns.
+Using the `CrossDomainMessenger` with a custom gas token would involve modifying the `sendMessage` implementation.
+
+### L1Block
+
+The `L1Block` contract is upgraded with a permissioned function that allows it to set the `GAS_PAYING_TOKEN_SLOT`.
+The `DEPOSITOR_ACCOUNT` MUST be the only account that can call the setter function. Any L2 contract that wants
+to learn the address of the gas paying token can call the getter on the `L1Block` contract.
+
+## User Flow
+
+Users should deposit native asset by calling `depositERC20Transaction` on the `OptimismPortal` contract.
+Users must first `approve` the address of the `OptimismPortal` so that the `OptimismPortal` can use
+`transferFrom` to take ownership of the ERC20 asset. If the `transferFrom` fails, a [permit2](https://github.com/Uniswap/permit2)
+`transferFrom` is attempted as a fallback.
+
+Users should withdraw value by calling the `L2ToL1MessagePasser` directly.
+
+## Upgrade
+
+The custom gas token upgrade is not yet defined to be part of a particular network upgrade, but it will be scheduled
+as part of a future hardfork. On the network upgrade block, a set of deposit transaction based upgrade transactions
+are deterministically generated by the derivation pipeline in the following order:
+
+- L1 Attributes Transaction calling `setL1BlockValuesEcotone`
+- User deposits from L1
+- Network Upgrade Transactions
+  - L1Block deployment
+  - L2CrossDomainMessenger deployment
+  - L2StandardBridge deployment
+  - Update L1Block Proxy ERC-1967 Implementation Slot
+  - Update L2CrossDomainMessenger Proxy ERC-1967 Implementation Slot
+  - Update L2StandardBridge Proxy ERC-1967 Implementation Slot
+
+### L1Block Deployment
+
+- `from`: `0x4210000000000000000000000000000000000002`
+- `to`: `null`
```

Comment:
This is true and @protolambda had the idea to fix this by defining every L2 contract upgrade via sets of upgrade txs. This would mean the `0xc0d3..` addresses would all be `v1.0.0` and then the genesis generation would apply all of the upgrade txs against the genesis state before spitting out a `genesis.json`. I consider this tech debt, we didn't have a good story for what L2 upgrades should be. To be honest I would like to make the L2 `ProxyAdmin` ownable by the `DEPOSITOR_ACCOUNT` and then define upgrade txs to go through there instead. We would have to have a common address space in the predeploys that we commit to never using that are free for application layer usage and they would need a diff `ProxyAdmin` to manage them

Reply:
Tracked in https://github.com/ethereum-optimism/protocol-quest/issues/222#issuecomment-2041822395